File: weather_api_test.py
import pygame
import requests

# OpenWeatherMap API key
api_key = 'bab5915f023f7655f2aa18da9b2dbe4e'

def get_weather_data_by_coordinates(latitude, longitude):
    base_url = 'http://api.openweathermap.org/data/2.5/weather'
    params = {'lat': latitude, 'lon': longitude, 'appid': api_key}

    try:
        response = requests.get(base_url, params=params)
        response.raise_for_status()  # Raise an HTTPError for bad responses
        data = response.json()
        return data
    except requests.exceptions.HTTPError as errh:
        print(f"HTTP Error: {errh}")
    except requests.exceptions.ConnectionError as errc:
        print(f"Error Connecting: {errc}")
    except requests.exceptions.Timeout as errt:
        print(f"Timeout Error: {errt}")
    except requests.exceptions.RequestException as err:
        print(f"Request Error: {err}")
    return None

def display_weather_data(data):
    if data and 'main' in data:
        print(data, "\n")  # Display the full data for reference

        # Extract relevant information from the API response
        temperature = data['main']['temp'] - 273
        weather_description = data['weather'][0]['description']

        # Display the weather information
        print(f'Temperature: {temperature}C')
        print(f'Weather: {weather_description}')

        # Graphics rendering based on weather
        set_background_color(weather_description)

def set_background_color(weather_description):
    # Set background color based on weather description
    if 'clear' in weather_description.lower():
        screen.fill((135, 206, 250))  # Light Blue for clear sky
    elif 'cloud' in weather_description.lower():
        screen.fill((192, 192, 192))  # Silver for cloudy
    elif 'rain' in weather_description.lower():
        screen.fill((0, 0, 139))  # Dark Blue for rain
    elif 'haze' in weather_description.lower():
        screen.fill((255, 255, 127))
    else:
        screen.fill((255, 255, 255))  # White for other conditions

def main():
    pygame.init()

    # Set up the graphical window
    window_size = (800, 600)
    global screen
    screen = pygame.display.set_mode(window_size)
    pygame.display.set_caption('Weather Visualization')

    running = True
    # Surabaya
    # latitude = -7.2575
    # longitude = 112.7521
    
    # Jeddah
    latitude = 21.492500
    longitude = 39.177570

    while running:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False

        # Fetch weather data
        weather_data = get_weather_data_by_coordinates(latitude, longitude)

        # Display weather data and update graphics
        display_weather_data(weather_data)

        pygame.display.flip()

    pygame.quit()

if __name__ == "__main__":
    main()

File: weather_api_test.js
const apiKey = 'bab5915f023f7655f2aa18da9b2dbe4e';

async function getWeatherDataByCoordinates(latitude, longitude) {
    const baseUrl = 'http://api.openweathermap.org/data/2.5/weather';
    const params = `lat=${latitude}&lon=${longitude}&appid=${apiKey}`;

    try {
        const response = await fetch(`${baseUrl}?${params}`);
        const data = await response.json();
        return data;
    } catch (error) {
        console.error('Error fetching weather data:', error);
        return null;
    }
}

function displayWeatherData(data) {
    if (data && 'main' in data) {
        console.log(data);  // Display the full data for reference

        // Extract relevant information from the API response
        const temperature = data.main.temp - 273;
        const weatherDescription = data.weather[0].description;

        // Display the weather information
        console.log(`Temperature: ${temperature}°C`);
        console.log(`Weather: ${weatherDescription}`);

        // Graphics rendering based on weather
        setBackgroundColor(weatherDescription);
    }
}

function setBackgroundColor(weatherDescription) {
    // Set background color based on weather description
    const canvas = document.getElementById('weatherCanvas');
    const context = canvas.getContext('2d');

    if (typeof weatherDescription === 'string') {
        if (weatherDescription.toLowerCase().includes('clear')) {
            context.fillStyle = 'lightblue';  // Light Blue for clear sky
        } else if (weatherDescription.toLowerCase().includes('cloud')) {
            context.fillStyle = 'silver';  // Silver for cloudy
        } else if (weatherDescription.toLowerCase().includes('rain')) {
            context.fillStyle = 'darkblue';  // Dark Blue for rain
        } else if (weatherDescription.toLowerCase().includes('haze')) {
            context.fillStyle = 'rgb(255, 255, 127)';
        } else {
            context.fillStyle = 'white';  // White for other conditions
        }
    } else {
        console.error('Invalid weather description:', weatherDescription);
        // Handle the case where weatherDescription is not a string
        context.fillStyle = 'white';  // Default to white for other conditions
    }

    context.fillRect(0, 0, canvas.width, canvas.height);
}

async function main() {
    // Set up the graphical window
    const canvas = document.createElement('canvas');
    canvas.id = 'weatherCanvas';
    canvas.width = 800;
    canvas.height = 600;
    document.body.appendChild(canvas);

    // Coordinates for Surabaya
    const latitude = -7.288131;
    const longitude = 112.802896;

    // Fetch weather data and update graphics
    const weatherData = await getWeatherDataByCoordinates(latitude, longitude);
    displayWeatherData(weatherData);

    // You may want to use setInterval or other techniques to update the weather regularly

    // Note: In a complete application, you would likely use a front-end framework like React, Angular, or Vue.js for a more structured approach.
}

if (typeof window !== 'undefined') {
    // Ensure the script is run only in a browser environment
    main();
}
